rfc/wallpapers_v2_network_improvements.md

```diff
@@ -0,0 +1,69 @@
+in allow for delivering wallpapers from a remote source. As part of a
+larger set of improvements planned for Q3 2022, it is intended to revisit
+this implementation to create a more scalable version.
+
+---
+
+## Motivation
+
+The original version of the networking solution consisted of
+adding wallpaper image files to deeply nested directories which defined the properties
+of the wallpaper. For example, an Android wallpaper might be found at:
+```
+<root>/mobile-wallpapers/android/<resolution>/<orientation>/<light/dark>/<wallpaper collection>/<wallpaper name>
+```
+
+This directory structure was confusing and difficult to manage. Copying several versions of a file to different locations
+was painful, even when automated. Automation would require consistent file naming for each new set of wallpapers, which can
+be hard to guarantee. 
+
+Downloads were started automatically during app startup if a wallpaper was missing.
+
+The files were also hosted by another Mozillian team, which may introduce complexity in terms of ownership or cost.
+
+In general, we would like to achieve the following goals:
+
+- it is easier for product, UX, and ENG to add new wallpapers
+- it is possible to determine important metadata about a wallpaper, like its size or average color
+- download is transparent to the user
+- download is activated by the user
+- require less bandwidth, if possible
+- hosted in a more permanent, team-controlled location
+
+---
+
+## Guide-level explanation
+
+Instead of relying on an (undocumented) convention for file paths, wallpaper metadata will be contained in a top-level JSON schema.
+Using a well-known format will help discoverability, and JSON's extensiblility should allow for simpler long-term maintenance. 
+It can also provide a more direct path to any required automation. This schema could define things like:
+
+- Android file path(s)
+- iOS file path(s)
+- thumbnail path(s)
+- average color (for text and background coloring purposes)
+- collection grouping for things like seasonal promotions
+- wallpaper expiration date
+- wallpaper region availability
+
+We will likely continue to ship some small number of wallpapers as part of our app binaries. This ensures that at least a few wallpapers are available for onboarding. We could also
+just ship a few thumbnails and require users to download the wallpaper they are interested in once they reach the wallpaper settings page.
+
+Additionally, including thumbnails will allow clients to keep download size relatively small. Thumbnails will need to follow the current eager downloading strategy
+that wallpapers currently do. This will allow thumbnails to be decoupled from app releases, but they will still be visible when the settings page is open. The current suggestion is to
+display some kind of action overlay above the thumbnails indicating that clicking them will cause a download.
+
+---
+
+## Drawbacks
+
+- Maintenance burden of owning the hosting solution
```

Comment:
There is a big drawback with the JSON approach - there is no schema that tells you how to read this JSON object.

Consider that we start with a structure like this:
```json
{
  "importantField": "value"
}
```
Code is released on app version 1.0 that knows how to read this structure.

We later need to support a list of values in the future, so one possible solution is:
```json
{
  "importantField": ["value", "anotherValue"]
}
```
Code is released on app version 2.0 that knows how to read this structure.

We have now divided our release users who are either on 1.0 or 2.0 of the code base, and the previous versions are no longer able to read the configuration.

If our argument is that we will take an abundance of care when adding new fields, that reasoning will only hold for so long and our parsing of this JSON code will get more and more complex over time.

---

I believe we should add another section called "Rationale and alternatives" to also consider other schematized structures or versioned paths to the JSON files.

We should also also some of the other teams you may have worked with for their input here since they will probably have run into this problem and solved for it already.

Reply:
These are all excellent points, thanks for the detailed feedback! Though I agree that's a risk, it could be mitigated by careful updates - adding new fields instead of updating the types of fields would avoid this.

> We should also also some of the other teams you may have worked with for their input here since they will probably have run into this problem and solved for it already.

Any suggestions for who to talk to? Christian has suggested eventually moving to `RemoteSettings`, but that will need to wait until both iOS and Android have adopted it.

> I believe we should add another section called "Rationale and alternatives" to also consider other schematized structures or versioned paths to the JSON files.

Versioned JSON schemas based on paths seem like a straightforward implementation for this concern. I've added some detail to the RFC suggesting as much, as well as the "Rationale and alternatives" section